planning/scenario_planning/common/motion_velocity_optimizer/src/motion_velocity_optimizer.cpp

```diff
@@ -661,18 +708,77 @@ void MotionVelocityOptimizer::updateExternalVelocityLimit(const double dt)
   }
 }
 
-void MotionVelocityOptimizer::timerCallback(const ros::TimerEvent & e)
+void MotionVelocityOptimizer::timerCallback()
 {
-  const double dt = (e.current_expected - e.last_expected).toSec();
+  const double dt = 1.0 / external_velocity_limit_update_rate_;
   updateExternalVelocityLimit(dt);
 }
 
-int main(int argc, char ** argv)
+void MotionVelocityOptimizer::blockUntilVehiclePositionAvailable(const tf2::Duration & timeout)
 {
-  ros::init(argc, argv, "motion_velocity_optimizer");
-  MotionVelocityOptimizer obj;
+  auto cti = std::make_shared<tf2_ros::CreateTimerROS>(
+    this->get_node_base_interface(), this->get_node_timers_interface());
+  tf_buffer_->setCreateTimerInterface(cti);
+
+  while (rclcpp::ok()) {
+    static constexpr auto input = "map", output = "base_link";
+    auto tf_future = tf_buffer_->waitForTransform(
+      input, output, tf2::TimePointZero, tf2::durationFromSec(0.0), [](auto &) {});
+    const auto status = tf_future.wait_for(timeout);
+    if (status == std::future_status::ready) {
+      break;
+    } else {
+      RCLCPP_INFO(
+        get_logger(), "waiting another %d seconds for %s->%s transform",
+        std::chrono::duration_cast<std::chrono::seconds>(timeout).count(), input, output);
+    }
+  }
+}
 
-  ros::spin();
+rcl_interfaces::msg::SetParametersResult MotionVelocityOptimizer::paramCallback(
+  const std::vector<rclcpp::Parameter> & parameters)
+{
+  rcl_interfaces::msg::SetParametersResult result;
+  result.successful = true;
+  result.reason = "success";
+
+  try {
+    UPDATE_PARAM(max_velocity);
+    UPDATE_PARAM(max_accel);
+    UPDATE_PARAM(min_decel);
+    UPDATE_PARAM(max_lateral_accel);
+    UPDATE_PARAM(min_curve_velocity);
+    UPDATE_PARAM(decel_distance_before_curve);
+    UPDATE_PARAM(decel_distance_after_curve);
+    UPDATE_PARAM(replan_vel_deviation);
+    UPDATE_PARAM(engage_velocity);
+    UPDATE_PARAM(engage_acceleration);
+    UPDATE_PARAM(engage_exit_ratio);
+    UPDATE_PARAM(extract_ahead_dist);
+    UPDATE_PARAM(extract_behind_dist);
+    UPDATE_PARAM(stop_dist_to_prohibit_engage);
+    UPDATE_PARAM(delta_yaw_threshold);
+
+    UPDATE_PARAM(resample_time);
+    UPDATE_PARAM(resample_dt);
+    UPDATE_PARAM(max_trajectory_length);
+    UPDATE_PARAM(min_trajectory_length);
+    UPDATE_PARAM(min_trajectory_interval_distance);
+
+    optimizer_->setAccel(planning_param_.max_accel);
+    optimizer_->setDecel(planning_param_.min_decel);
+  } catch (const rclcpp::exceptions::InvalidParameterTypeException & e) {
+    result.successful = false;
+    result.reason = e.what();
+  }
+
+  return result;
+}
 
+int main(int argc, char ** argv)
```

Comment:
It would be cleaner to have the main function in a separate file

Reply:
Done